echopype/preprocess/api.py

```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+        2D array of bin indices for ``echo_range``
+    n_bin_er: int
+        The number of echo range bins
+
+    Returns
+    -------
+    er_means: dask.array.Array or np.ndarray
+        The bin and mean of ``arr`` with respect to ``echo_range``
+    """
+
+    # compute bin indices to allow for downstream processes (mainly axis argument in unique)
+    if isinstance(digitized_echo_range, dask.array.Array):
+        digitized_echo_range = digitized_echo_range.compute()
+
+    # determine the unique rows of digitized_echo_range and the inverse
+    unique_er_bin_ind, unique_inverse = np.unique(digitized_echo_range, axis=0, return_inverse=True)
+
+    # create groups of row indices using the unique inverse
+    grps_same_ind = [
+        np.argwhere(unique_inverse == grp).flatten() for grp in np.unique(unique_inverse)
+    ]
+
+    # for each group bin and mean arr along echo_range
+    # note: the values appended may not be in the correct final order
+    binned_er = []
+    for count, grp in enumerate(grps_same_ind):
+        binned_er.append(
+            bin_and_mean_echo_range(arr[grp, :], unique_er_bin_ind[count, :], n_bin_er)
+        )
+
+    # construct er_means and put the columns in the correct order
+    binned_er_array = np.hstack(binned_er)
+    correct_column_ind = np.argsort(np.concatenate(grps_same_ind))
+    er_means = binned_er_array[:, correct_column_ind]
+
+    return er_means
+
+
+def bin_and_mean_2d(
+    arr: Union[dask.array.Array, np.ndarray],
+    bins_time: np.ndarray,
+    bins_er: np.ndarray,
+    times: np.ndarray,
+    echo_range: np.ndarray,
+    comp_er_check: bool = True,
+) -> np.ndarray:
+    """
+    Bins and means ``arr`` based on ``times`` and ``echo_range``,
+    and their corresponding bins. If ``arr`` is ``Sv`` then this
+    will compute the MVBS.
+
+    Parameters
+    ----------
+    arr: dask.array.Array or np.ndarray
+        The 2D array whose values should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    echo_range: np.ndarray
+        2D array of echo range values
+    comp_er_check: bool
+        If True, a more comprehensive check will be completed to determine if ``echo_range``
+        grouping along ``ping_time`` is needed, otherwise a less comprehensive check will be done
+
+    Returns
+    -------
+    final_reduced: np.ndarray
+        The final binned and mean ``arr``, if ``arr`` is ``Sv`` then this is the MVBS
+
+    Notes
+    -----
+    This function assumes that ``arr`` has rows corresponding to
+    ``ping_time`` and columns corresponding to ``echo_range``.
+
+    This function should not be run if the number of ``echo_range`` values
+    vary amongst ``ping_times``.
```

Comment:
By "the number of `echo_range` values" do you mean the number of bins along echo range? In the case of data created by echopype, the scenario of the number of `echo_range` values varying across `ping_times` would never occur since we NaN-padded the shorter pings, right?

Reply:
> By "the number of echo_range values" do you mean the number of bins along echo range?

By the "the number of echo_range values" I meant that the 2D array `echo_range` should have columns that are of the same length (i.e. no ragged arrays). 

> In the case of data created by echopype, the scenario of the number of echo_range values varying across ping_times would never occur since we NaN-padded the shorter pings, right?

Yes, this should never occur for `compute_MVBS` because we padded the shorter pings. I put this comment there just incase in the future we wanted to use this function for a different application.

Reply:
Ok I see. Could you please also add this second part (that this should not occur for our current use of echopype-generated Sv data) so that it is clear what this is intended for? Thanks!

Reply:
No problem, I will go ahead and add a comment here.